2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/sinks/PubSubSink.scala

```diff
@@ -0,0 +1,175 @@
+      .setInitialRetryDelay(Duration.ofMillis(minBackoff))
+      .setRetryDelayMultiplier(retryDelayMultiplier)
+      .setTotalTimeout(Duration.ofMillis(totalTimeOut))
+      .setInitialRpcTimeout(Duration.ofMillis(initialRpcTimeout))
+      .setMaxRpcTimeout(Duration.ofMillis(maxRpcTimeout))
+      .build
+
+  /**
+   * Checks if a given pubsub topic exists
+   *
+   * @return Boolean
+   */
+  private def checkTopicExists(topic: TopicName): Boolean = {
+    val topicAdminClient = TopicAdminClient.create()
+    try {
+      val response = topicAdminClient.getTopic(topic)
+      true
+    } catch {
+      case e: com.google.api.gax.grpc.GrpcApiException => {
+        if (e.getStatusCode().getCode() == Status.Code.NOT_FOUND) {
+          false
+        } else {
+          false
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks if a topic has >0 subscriptions attached
+   *
+   * @return Boolean
+   */
+  private def hasSubscriptions(topic: TopicName): Boolean = {
+    val subscriptionAdminClient = SubscriptionAdminClient.create()
+    val topicAdminClient = TopicAdminClient.create()
+    val topicSubscriptionsRequest = ListTopicSubscriptionsRequest.newBuilder()
+      .setTopicWithTopicName(topic)
+      .build()
+
+    val response = topicAdminClient.listTopicSubscriptions(topicSubscriptionsRequest);
+
+    val subscriptions = response.iterateAll()
+    if (subscriptions.size > 0) true else false
+  }
+
+  /**
+   * Instantiates a Publisher on an existing topic
+   * with the given configuration options. If the name isn't correct, this will fail
+   *
+   * @return a PubSub publisher
+   */
+  private def createPublisher: Publisher = {
+    val topic = TopicName.create(s"${config.googleProjectId}", s"$topicName")
+    val topicExists = checkTopicExists(topic)
+    if (topicExists == false) {
+      throw new RuntimeException(s"The pubsub topic $topicName was not found.")
+    }
+    // val topicHasSubscriptions = hasSubscriptions(topic)
+    // if (topicHasSubscriptions == false) {
+    //   warn("The topic $topicName has no associated subscriptions")
+    // }
+    val publisher = Publisher.defaultBuilder(
+        topic
+      )
+      .setBatchingSettings(batchingSettings)
+      .setRetrySettings(retrySettings)
+      .build
+    publisher
+  }
+
+  /**
+   * Convert event bytes to PubsubMessage to be published
+   * @param event Event to be converted
+   * @return PubsubMessage instance
+   */
+  private def eventToPubsubMessage(event: Array[Byte]): PubsubMessage = {
+    PubsubMessage.newBuilder
+      .setData(ByteString.copyFrom(event))
+      .build
+  }
+
+  /**
+   * Store raw events to the topic
+   *
+   * @param events The list of events to send
+   * @param key Not used.
+   */
+  override def storeRawEvents(events: List[Array[Byte]], key: String) = {
+    debug(s"Writing ${events.size} Thrift records to PubSub topic ${topicName}")
+    events.foreach(event => pubSubPublisher.publish(eventToPubsubMessage(event)))
```

Comment:
can't we extract something from `publish` to know whether or not the write succeeded?